### DISCUSSION TITLE
Lens contamination effects

### DISCUSSION THREAD

**Discussion by PhilippAuburger** [carla-simulator/carla#8121] [Q&A]
Dear all,

I want to simulate lens contamination effects (rain/dust/spray on the camera lens) using the CARLA simulator (0.9.14 or newer).
Can anyone give me an example how to do this (directly or via postprocessing)?

**Comment by Fillipko** [comment]
I found a couple of things that effect the camera lens: the `enable_post_process_effects` [camera attribute](https://carla.readthedocs.io/en/latest/ref_sensors/#rgb-camera) and the `wetness` [weather parameter](https://carla.readthedocs.io/en/latest/python_api/#carla.WeatherParameters). However, neither of these seem to add dust to the lens.

**Comment by Fillipko** [comment]
Aha! After digging into the documentation and some testing of my own, it looks like there are two `WeatherParameters` that effect the lens contamination: `precipitation` and `dust_storm`. Setting these to max intensity gets both water and dust on the camera respectively. Here's a simple example for setting it up:
```python
import carla

client = carla.Client("localhost", 2000)
client.set_timeout(10.0)
world = client.get_world()
weather = world.get_weather()

weather.precipitation = 100 # Change to decrease water intensity on lens
weather.dust_storm = 100 # Change to decrease dust intensity on lens

print("Changing the weather...")
world.set_weather(weather)
```

**Reply by Fillipko** [reply]
It would be cool to see an option for fogging up the glass to simulate colder temperatures or having water be splashed onto the lens from another vehicle, but those would probably have to be future developments.
